daemon/api_aspects.go

```diff
@@ -0,0 +1,118 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/snapcore/snapd/aspects"
+	"github.com/snapcore/snapd/overlord/auth"
+)
+
+var (
+	aspectsCmd = &Command{
+		Path:        "/v2/aspects",
+		GET:         getAspect,
+		POST:        setAspect,
+		ReadAccess:  openAccess{},
+		WriteAccess: authenticatedAccess{Polkit: polkitActionManage},
+	}
+)
+
+type AspectRequest struct {
+	Account    string      `json:"account"`
+	BundleName string      `json:"bundle"`
+	Aspect     string      `json:"aspect"`
+	Field      string      `json:"field"`
+	Value      interface{} `json:"value"`
+}
+
+const emptyFieldFmt = "cannot have empty %q field"
+
+func (r *AspectRequest) validate() error {
+	if r.Account == "" {
+		return fmt.Errorf(emptyFieldFmt, "account")
+	} else if r.BundleName == "" {
+		return fmt.Errorf(emptyFieldFmt, "bundle")
+	} else if r.Aspect == "" {
+		return fmt.Errorf(emptyFieldFmt, "aspect")
+	} else if r.Field == "" {
+		return fmt.Errorf(emptyFieldFmt, "field")
+	}
+
+	return nil
+}
+
+func getAspect(c *Command, r *http.Request, _ *auth.UserState) Response {
+	decoder := json.NewDecoder(r.Body)
+	var req AspectRequest
+	if err := decoder.Decode(&req); err != nil {
+		return BadRequest("cannot decode aspect request body: %v", err)
+	}
+
+	if err := req.validate(); err != nil {
+		return BadRequest(err.Error())
+	}
+
+	var value interface{}
+	err := aspectstateGet(c.d.state, req.Account, req.BundleName, req.Aspect, req.Field, &value)
+	if err != nil {
+		if aspects.IsNotFoundErr(err) {
+			return NotFound(err.Error())
+		} else if errors.Is(err, &aspects.InvalidAccessError{}) {
+			return &apiError{
+				Status:  403,
+				Message: err.Error(),
+			}
+		}
+		return InternalError(err.Error())
+	}
+
+	return SyncResponse(value)
+}
+
+func setAspect(c *Command, r *http.Request, _ *auth.UserState) Response {
+	decoder := json.NewDecoder(r.Body)
```

Comment:
It looks like there should be checks for content /media / charset ? (perhaps something similar to [this](https://github.com/snapcore/snapd/blob/master/daemon/api_snaps.go#L605) )

Reply:
The snippet you linked uses the media type to differentiate between a store install and a local install, it's not quite a validation. The charset isn't required to be there either, it's just required not to be explicitly wrong (and apparently only for the `postSnaps` case). There also aren't any checks in `postSnap`, `getSnapInfo` and `getSnapsInfo` AFAICT. So I think we just accept requests with valid JSON carrying the required fields.